File: 0139-word-break/0139-word-break.cpp
class Solution {
public:
    int ns,nd;
    unordered_set<string>st;
    bool fun(int i,string s,vector<int>&dp){
        if(i>=ns) return true;
        if(dp[i]!=-1) return dp[i];
        for(int idx=i;idx<ns;idx++){
            string x=s.substr(i,idx-i+1);
            if(st.count(x)){
              if(fun(idx+1,s,dp)) return dp[i]=true;
            }
        }
        return dp[i]=false;
    
    }
    bool wordBreak(string s, vector<string>& wordDict) {
        ns=s.size(),nd=wordDict.size();
        for(int i=0;i<nd;i++){
            st.insert(wordDict[i]);
        }
        vector<int>dp(ns,-1);
        return fun(0,s,dp);

    }
};

File: 0139-word-break/0139-word-break.py
class Solution:
    def fun(self, start, s, wordDict,dp):
        if start == len(s):
            return True
        if dp[start]!=-1 :
            return dp[start]

        for i in range(start, len(s)):
            x = s[start:i+1]

            if x in wordDict:
                if self.fun(i+1, s, wordDict,dp):
                    dp[i]=True
                    return dp[i]

        dp[start]=False
        return dp[start]


    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        st = set(wordDict)
        dp=[-1]*(len(s)+1)

        return self.fun(0, s, st,dp)